internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +114,183 @@ func New(ctx context.Context,
 //
 // Cancel is an optional function to release resources in the producer. If
 // provided, the consumer may call this function multiple times.
-func (c *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
-	// TODO implement.
-	return nil, nil
+func (o *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
+	ctx := context.Background()
+
+	// If this is the first time we're seeing the request from the
+	// downstream client, initialize a channel to feed future responses.
+	responseChannel := o.downstreamResponseMap.createChannel(&req)
+
+	aggregatedKey, err := o.mapper.GetKey(req)
+	if err != nil {
+		// Can't map the request to an aggregated key. Log and continue to
+		// propagate the response upstream without aggregation.
+		o.logger.With("err", err).With("req node", req.GetNode()).Warn(ctx, "failed to map to aggregated key")
+		// Mimic the aggregated key.
+		// TODO (https://github.com/envoyproxy/xds-relay/issues/56). This key
+		// needs to be made more granular to uniquely identify a request.
+		aggregatedKey = fmt.Sprintf("%s%s_%s", unaggregatedPrefix, req.GetNode().GetId(), req.GetTypeUrl())
+	}
+
+	// Register the watch for future responses.
+	err = o.cache.AddRequest(aggregatedKey, &req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req node", req.GetNode()).Error(ctx, "failed to add watch")
+		closedChannel := o.downstreamResponseMap.delete(&req)
+		return closedChannel, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Warn(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil && cached.Resp.Raw.GetVersionInfo() != req.GetVersionInfo() {
+		// If we have a cached response and the version is different,
+		// immediately push the result to the response channel.
+		go func() { responseChannel <- convertToGcpResponse(cached.Resp, req) }()
+	}
+
+	// Check if we have a upstream stream open for this aggregated key. If not,
+	// open a stream with the representative request.
+	//
+	// Locking is necessary here so that a simultaneous downstream request
+	// that maps to the same aggregated key doesn't result in two upstream
+	// streams.
+	o.upstreamResponseMap.mu.Lock()
```

Comment:
can we use a more granular lock per stream/key?

Reply:
I don't think we can use a more granular lock because we're modifying the map to add keys further down https://github.com/envoyproxy/xds-relay/pull/53/files#diff-2958638164db92742469d2bb08d0001aR173

Reply:
All simultaneous requests will be blocking on the slowest OpenStream, isnt it?

Reply:
I moved to `sync.Map` for per-key concurrency, let me know your thoughts: https://github.com/envoyproxy/xds-relay/pull/53/commits/aeef7ca998b42f1d2fe9237d17186003a397f91b